numba/cgutils.py

```diff
@@ -657,16 +657,62 @@ def unpack_tuple(builder, tup, count=None):
     return vals
 
 
-def get_item_pointer(builder, aryty, ary, inds, wraparound=False):
+def get_item_pointer(context, builder, aryty, ary, inds, wraparound=False,
+                     boundscheck=False):
+    # Set boundscheck=True for any pointer access that should be
+    # boundschecked. do_boundscheck() will handle enabling or disabling the
+    # actual boundschecking based on the user config.
     shapes = unpack_tuple(builder, ary.shape, count=aryty.ndim)
     strides = unpack_tuple(builder, ary.strides, count=aryty.ndim)
-    return get_item_pointer2(builder, data=ary.data, shape=shapes,
+    return get_item_pointer2(context, builder, data=ary.data, shape=shapes,
                              strides=strides, layout=aryty.layout, inds=inds,
-                             wraparound=wraparound)
-
-
-def get_item_pointer2(builder, data, shape, strides, layout, inds,
-                      wraparound=False):
+                             wraparound=wraparound, boundscheck=boundscheck)
+
+
+def do_boundscheck(context, builder, ind, dimlen, axis=None):
+    # Boundschecking is always disabled for CUDA
+    from .cuda.target import CUDATargetContext
```

Comment:
This is odd. Can we rely on turning off boundchecking at the `numba/cuda/compiler.py`.

And it is likely causing other errors because the `cuda` subpackage is auto-initializing the CUDA Driver.

Reply:
The reason I have it this way is because I have it so that the environment variable overrides the flag.

Reply:
This import is causing error like: https://dev.azure.com/numba/numba/_build/results?buildId=3349&view=logs&jobId=eb149424-352d-52d9-b16d-aeffaad6858a&taskId=cdac5993-002e-5764-1ca7-e646ebe78687&lineStart=1728&lineEnd=1760&colStart=1&colEnd=99

Reply:
OK I didn't realize that the import could fail like that. What would be the best way to fix this?

Reply:
Can we add some attribute to the context that lets us detect that it is CUDA without importing it? For instance, `context.name`. Or I can just check `type(context).__name__`.

Reply:
I don't think this function should do the check at all.  If we turn off `boundscheck` in cuda context (https://github.com/numba/numba/pull/4432#discussion_r351353479) and make sure the array implementation follow the context setting, we can avoid the check here.   

Here's what I check the changes should look like: 
https://gist.github.com/sklam/38a31d9a5e9124dab7a77def216e0a4c
(I haven't fully tested it)

Reply:
I'll check it out. Actually this reminds me that I was going to add a test that the NUMBA_BOUNDSCHECK environment variable overrides the flag correctly.

Reply:
I had to modify it so that the `NUMBA_BOUNDSCHECK` environment variable still works. I did this by making `enable_boundscheck` a property on the context class, so that it can use the config.BOUNDSCHECK as a default and so it can be modified at runtime by the `override_env_config` context manager for testing.